### PR TITLE
Metal API prototype.

### PR DESCRIPTION
**Pull request by Rsedaikin**
Just first prototype of Metal support. Bug fixes needed.

### CODE REVIEW

File: .project [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (olonho) [inline .project:1]: Unlikely it is needed.

File: skiko/.project [hunk 2]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (olonho) [inline skiko/.project:1]: Unlikely it is needed.

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt [hunk 3]
```diff
@@ -20,11 +22,47 @@ private class PictureHolder(val instance: Picture, val width: Int, val height: I
 
 open class SkiaLayer(
     private val properties: SkiaLayerProperties = SkiaLayerProperties()
-) : HardwareLayer() {
+) : JPanel() {
+
+    val backedLayer : HardwareLayer
```
Comment (olonho) [inline skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt:27]: Why not combine declaration and init?

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt [hunk 4]
```diff
@@ -20,11 +22,47 @@ private class PictureHolder(val instance: Picture, val width: Int, val height: I
...
+
+    init {
+        setOpaque(false)
```
Comment (olonho) [inline skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt:30]: Why cannot we use setter?

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt [hunk 5]
```diff
@@ -20,11 +22,47 @@ private class PictureHolder(val instance: Picture, val width: Int, val height: I
...
+        @Suppress("LeakingThis")
+        addHierarchyListener {
+            if (it.changeFlags and HierarchyEvent.SHOWING_CHANGED.toLong() != 0L) {
+                checkIsShowing()
+            }
+        }
+    }
+
+    private var isInit = false
+
```
Comment (olonho) [inline skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt:42]: isInited?

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/context/MetalContextHandler.kt [hunk 6]
```diff
@@ -30,16 +30,22 @@ internal class MetalContextHandler(layer: SkiaLayer) : ContextHandler(layer) {
...
 
-        renderTarget = makeMetalRenderTarget(w, h, 0)
+        renderTarget = metalRedrawer.makeRenderTarget(w, h)
 
         surface = Surface.makeFromBackendRenderTarget(
             context!!,
             renderTarget!!,
-            SurfaceOrigin.BOTTOM_LEFT,
-            SurfaceColorFormat.RGBA_8888,
+            SurfaceOrigin.TOP_LEFT,
```
Comment (olonho) [inline skiko/src/jvmMain/kotlin/org/jetbrains/skiko/context/MetalContextHandler.kt:38]: Why we select particular constant?

File: .project (hunk 1 above)
Comment (Rsedaikin) [reply .project]: Done.

File: skiko/.project (hunk 2 above)
Comment (Rsedaikin) [reply skiko/.project]: Done.

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt (hunk 3 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt:28]: Done.

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt (hunk 5 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt:49]: Done.

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt (hunk 4 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/kotlin/org/jetbrains/skiko/SkiaLayer.kt:36]: JComponent has "setOpaque" and "isOpaque" methods, so we cannot use "opaque" as setter/getter.

**Comment by olonho** [comment]
Looks good, please rebase.

**Review by olonho — APPROVED** [review]

**Pull request merged**